File: src/shared/develop/crop-aspect.ts
/**
 * What a crop's ASPECT is — the vocabulary a roll stores in
 * `RollPicture.aspect`, and the one place that turns it into a ratio.
 *
 * Three kinds of value on one string field:
 *
 *  - `'original'` — the picture's own shape, whatever it was shot at.
 *  - a preset id from `ASPECT_PRESETS` (`'4:5'`, `'16:9'`, …) — a destination's
 *    shape, the vocabulary the Studio and Trips compose for.
 *  - `'free:<ratio>'` — a FREE zone: any shape the author drew, carried as its
 *    own w/h.
 *
 * The free zone rides the SAME field rather than a second one, so every reader
 * that already asks for a ratio (`roll-render`, the export, the filmstrip
 * cell, the crop stage) keeps working unchanged, the batch verb copies a free
 * shape with no new case, and a roll written before free crops existed still
 * reads. What it costs is this module: one place that knows the spelling.
 *
 * Pure and DOM-free. The crop stage's arithmetic — what a handle, a move or a
 * rotation does to the zone — is `crop-rect.ts`.
 */

import { ASPECT_PRESETS } from '../projects/project-types';

const FREE_PREFIX = 'free:';

/**
 * The shapes a free crop is held between. Past 5:1 a photograph is a strip:
 * the frame would be a few pixels tall on the stage and there would be nothing
 * left to judge, so the slider and the corners both stop here.
 */
export const FREE_ASPECT_MAX = 5;
export const FREE_ASPECT_MIN = 1 / FREE_ASPECT_MAX;

export function clampFreeAspect(ratio: number): number {
  if (!Number.isFinite(ratio) || ratio <= 0) return 1;
  return Math.min(FREE_ASPECT_MAX, Math.max(FREE_ASPECT_MIN, ratio));
}

/**
 * `1.37209` → `'free:1.3721'`. Four decimals: finer than a pixel on a 4096px
 * edge, and short enough that the stored roll stays readable.
 */
export function freeAspectId(ratio: number): string {
  return `${FREE_PREFIX}${Number(clampFreeAspect(ratio).toFixed(4))}`;
}

/** The ratio a free aspect carries, or null when this id is not one. */
export function freeAspectRatio(aspect: string): number | null {
  if (!aspect.startsWith(FREE_PREFIX)) return null;
  const ratio = Number(aspect.slice(FREE_PREFIX.length));
  return Number.isFinite(ratio) && ratio > 0 ? clampFreeAspect(ratio) : null;
}

export function isFreeAspect(aspect: string): boolean {
  return freeAspectRatio(aspect) !== null;
}

const PRESET_IDS: ReadonlySet<string> = new Set(ASPECT_PRESETS.map((a) => a.id));

/** Whether a stored string is an aspect at all — what `readRollDoc` trusts. */
export function isStoredAspect(aspect: string): boolean {
  return aspect === 'original' || PRESET_IDS.has(aspect) || isFreeAspect(aspect);
}

/**
 * The w/h ratio a picture's crop frames into: a free zone's own shape, one of
 * the suite's aspect presets, or the picture's OWN shape while its aspect is
 * `'original'` (and a safe square before a source has decoded at all).
 */
export function pictureAspectRatio(aspect: string, sourceW: number, sourceH: number): number {
  const free = freeAspectRatio(aspect);
  if (free !== null) return free;
  if (aspect !== 'original') {
    const preset = ASPECT_PRESETS.find((p) => p.id === aspect);
    if (preset) return preset.w / preset.h;
  }
  return sourceW > 0 && sourceH > 0 ? sourceW / sourceH : 1;
}

/** A ratio said the way a shape is read: `1.50:1`, `1:1.25`, `1:1`. */
export function describeAspect(ratio: number): string {
  const r = clampFreeAspect(ratio);
  if (Math.abs(r - 1) < 0.005) return '1:1';
  return r > 1 ? `${r.toFixed(2)}:1` : `1:${(1 / r).toFixed(2)}`;
}

/** The format chip on screen: Free, Original, or one of the preset ids. */
export type CropChip = 'free' | 'original' | string;

/**
 * Which chip a picture OPENS on.
 *
 * An untouched picture opens on **Free** (2026-09-21, the maintainer's ask):
 * the Crop tab is reached in order to draw a shape, and opening on the
 * picture's own ratio meant every crop began with a click that undid the
 * default. Free is the one chip that holds the zone to nothing, so it is the
 * one that costs no gesture.
 *
 * It changes no stored value: Free only says what the next gesture MAY do, so
 * a picture opened on it and left alone still stores `'original'` with an
 * untouched framing, and `rollProgress` still counts it as uncropped.
 *
 * A picture that was already cropped opens on the chip its stored crop names
 * — a locked ratio comes back locked, `'original'` included.
 */
export function openingCropChip(aspect: string, untouched: boolean): CropChip {
  if (untouched) return 'free';
  if (aspect === 'original') return 'original';
  if (isFreeAspect(aspect)) return 'free';
  return PRESET_IDS.has(aspect) ? aspect : 'free';
}

export type CropHandle = 'n' | 's' | 'e' | 'w' | 'ne' | 'nw' | 'se' | 'sw';

File: apple/Packages/AtelierKit/Sources/AtelierKit/Develop/CropAspect.swift
// What a crop's ASPECT is — the vocabulary a roll stores in
// `RollPicture.aspect`, and the one place that turns it into a ratio. Port of
// `src/shared/develop/crop-aspect.ts`.
//
// Three kinds of value on one string field:
//  - `original` — the picture's own shape, whatever it was shot at;
//  - a preset id from `aspectPresets` (`4:5`, `16:9`, …) — a destination's shape;
//  - `free:<ratio>` — a FREE zone: any shape the author drew, carried as its own w/h.
//
// The free zone rides the SAME field rather than a second one, so every reader
// that asks for a ratio keeps working unchanged and a roll written before free
// crops existed still reads. What it costs is this module: one place that knows
// the spelling. The spelling is the web's to the character — four decimals, a
// whole number with none (`free:2`) — so a roll written here reads there.
//
// `isStoredAspect` lives here (it did in `Roll.swift`, without the free zone,
// which made a free crop read back as `original` and lose its shape).

import Foundation

private let freePrefix = "free:"

/// The shapes a free crop is held between. Past 5:1 a photograph is a strip:
/// the frame would be a few pixels tall on the stage and there would be nothing
/// left to judge, so the slider and the corners both stop here.
public let freeAspectMax = 5.0
public let freeAspectMin = 1 / freeAspectMax

public func clampFreeAspect(_ ratio: Double) -> Double {
    if !ratio.isFinite || ratio <= 0 { return 1 }
    return min(freeAspectMax, max(freeAspectMin, ratio))
}

/// `1.37209` → `free:1.3721`. Four decimals: finer than a pixel on a 4096 px
/// edge, and short enough that the stored roll stays readable.
public func freeAspectId(_ ratio: Double) -> String {
    let fixed = String(format: "%.4f", clampFreeAspect(ratio))
    return freePrefix + jsNumberString(Double(fixed) ?? 1)
}

/// The ratio a free aspect carries, or nil when this id is not one.
public func freeAspectRatio(_ aspect: String) -> Double? {
    guard aspect.hasPrefix(freePrefix) else { return nil }
    let ratio = jsNumber(String(aspect.dropFirst(freePrefix.count)))
    return ratio.isFinite && ratio > 0 ? clampFreeAspect(ratio) : nil
}

public func isFreeAspect(_ aspect: String) -> Bool {
    freeAspectRatio(aspect) != nil
}

/// Whether a stored string is an aspect at all — what `readRollDoc` trusts.
public func isStoredAspect(_ aspect: String) -> Bool {
    aspect == "original" || aspectPreset(aspect) != nil || isFreeAspect(aspect)
}

/// The w/h ratio a picture's crop frames into: a free zone's own shape, one of
/// the suite's aspect presets, or the picture's OWN shape while its aspect is
/// `original` (and a safe square before a source has decoded at all).
public func pictureAspectRatio(_ aspect: String, _ sourceW: Double, _ sourceH: Double) -> Double {
    if let free = freeAspectRatio(aspect) { return free }
    if aspect != "original", let preset = aspectPreset(aspect) { return preset.w / preset.h }
    return sourceW > 0 && sourceH > 0 ? sourceW / sourceH : 1
}

/// A ratio said the way a shape is read: `1.50:1`, `1:1.25`, `1:1`.
public func describeAspect(_ ratio: Double) -> String {
    let r = clampFreeAspect(ratio)
    if abs(r - 1) < 0.005 { return "1:1" }
    return r > 1 ? "\(String(format: "%.2f", r)):1" : "1:\(String(format: "%.2f", 1 / r))"
}

/// The format chip on screen: `free`, `original`, or one of the preset ids.
public typealias CropChip = String

/// Which chip a picture OPENS on.
///
/// An untouched picture opens on **Free** (2026-09-21, the maintainer's ask):
/// the Crop tab is reached in order to draw a shape, and Free is the one chip
/// that holds the zone to nothing. It changes no stored value — a picture
/// opened on it and left alone still stores `original` with an untouched
/// framing. A picture already cropped opens on the chip its stored crop names.
public func openingCropChip(_ aspect: String, untouched: Bool) -> CropChip {
    if untouched { return "free" }
    if aspect == "original" { return "original" }
    if isFreeAspect(aspect) { return "free" }
    return aspectPreset(aspect) != nil ? aspect : "free"
}

/// A handle of the crop zone: an edge or a corner, by compass point.
public enum CropHandle: String, CaseIterable, Sendable {
    case n, s, e, w, ne, nw, se, sw
}

// MARK: - JavaScript's number spelling

/// JavaScript's `Number(string)` on the strings a roll can hold: surrounding
/// white space ignored, the empty string 0, a decimal literal (sign, digits, a
/// point, an exponent) read as one, `Infinity` spelled out, and anything else
/// — `banana`, `1.5x`, Swift's own `inf` or hex floats — NaN.
private func jsNumber(_ text: String) -> Double {
    let s = text.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.isEmpty { return 0 }
    switch s {
    case "Infinity", "+Infinity": return .infinity
    case "-Infinity": return -.infinity
    default: break
    }
    var digits = 0
    var sawPoint = false
    var sawExponent = false
    var exponentDigits = 0
    var index = s.startIndex
    if s[index] == "+" || s[index] == "-" { index = s.index(after: index) }
    while index < s.endIndex {
        let c = s[index]
        if c.isASCII, c.isNumber {
            if sawExponent { exponentDigits += 1 } else { digits += 1 }
        } else if c == "." && !sawPoint && !sawExponent {
            sawPoint = true
        } else if (c == "e" || c == "E") && !sawExponent && digits > 0 {
            sawExponent = true
            let next = s.index(after: index)
            if next < s.endIndex, s[next] == "+" || s[next] == "-" { index = next }
        } else {
            return .nan
        }
        index = s.index(after: index)
    }
    if digits == 0 || (sawExponent && exponentDigits == 0) { return .nan }
    return Double(s) ?? .nan
}

/// A number the way JavaScript prints it in a template string, for the values
/// an aspect id holds: a whole number with no decimals, else the shortest
/// spelling that reads back as the same double.
private func jsNumberString(_ v: Double) -> String {
    if v == v.rounded(), abs(v) < 1e15 { return String(Int64(v)) }
    return "\(v)"
}
